diff --git a/tigerlings/003_create_account.sh b/tigerlings/003_create_account.sh
index 6a8a380a2..92825561c 100755
--- a/tigerlings/003_create_account.sh
+++ b/tigerlings/003_create_account.sh
@@ -5,7 +5,7 @@
 # To connect to TigerBeetle, we normally use one of the client libraries.
 # However, for these exercises we're just going to use the REPL command via the CLI.
 
-./tigerbeetle repl --cluster=0 --addresses=3000 --command="create_account id=300 code=10 ledger=30"
+./tigerbeetle repl --cluster=0 --addresses=3000 --command="create_accounts id=300 code=10 ledger=30"
 # Uh oh! This command isn't quite right. In TigerBeetle, every request can be batched for 
 # performance so the type of request is always plural. Here, it should be `create_accounts`!
 
