diff --git a/tigerlings/006_ledgers.sh b/tigerlings/006_ledgers.sh
index 6d4c3008a..9c8af6997 100644
--- a/tigerlings/006_ledgers.sh
+++ b/tigerlings/006_ledgers.sh
@@ -9,7 +9,7 @@ tb "create_accounts id=600 code=10 ledger=60,
                     id=602 code=10 ledger=61;"
 
 # And we'll create a transfer between them:
-tb "create_transfers id=60000 debit_account_id=600 credit_account_id=602 amount=100;"
+tb "create_transfers id=60000 debit_account_id=601 credit_account_id=602 amount=100;"
 # Uh oh! Only accounts on the same ledger can transfer directly to one another.
 # Try modifying the command to send the transfer between the two accounts on the same ledger.
 
